File: lambda-functions/assign-categorizer.js
// AWS Lambda Function - Assign Categorizer (Node.js)
// Matches your existing eventregistrations table structure
const { DynamoDBClient, ScanCommand } = require("@aws-sdk/client-dynamodb");
const { DynamoDBDocumentClient, PutCommand, UpdateCommand, QueryCommand, DeleteCommand } = require("@aws-sdk/lib-dynamodb");

// Initialize DynamoDB client
const REGION = "ap-south-1";
const client = new DynamoDBClient({ region: REGION });
const docClient = DynamoDBDocumentClient.from(client);

// Your existing table name
const TABLE_NAME = 'eventregistrations';
// New table for categorizer assignments
const CATEGORIZER_ASSIGNMENTS_TABLE = 'categorizer-assignments';

const corsHeaders = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'POST, OPTIONS'
};

exports.handler = async (event) => {
    try {
        // Handle OPTIONS for CORS preflight
        if (event.requestContext?.http?.method === 'OPTIONS' || event.httpMethod === 'OPTIONS') {
            return {
                statusCode: 200,
                headers: corsHeaders,
                body: ''
            };
        }

        // Parse request body
        const body = typeof event.body === 'string' ? JSON.parse(event.body) : event;

        // Support both old (registrationIds) and new (candidates) formats for backward compatibility
        let candidatesToProcess = [];
        let totalCount = 0;

        if (body.candidates && Array.isArray(body.candidates)) {
            candidatesToProcess = body.candidates;
            totalCount = candidatesToProcess.length;
        } else if (body.registrationIds && Array.isArray(body.registrationIds)) {
            console.warn('Received old format registrationIds, this might be slow');
            candidatesToProcess = body.registrationIds.map(id => ({ registrationId: id }));
            totalCount = candidatesToProcess.length;
        }

        const { categorizerId, categorizerName } = body;

        if (totalCount === 0) {
            return {
                statusCode: 400,
                headers: corsHeaders,
                body: JSON.stringify({
                    success: false,
                    message: 'candidates array or registrationIds array is required'
                })
            };
        }

        if (!categorizerId) {
            return {
                statusCode: 400,
                headers: corsHeaders,
                body: JSON.stringify({
                    success: false,
                    message: 'categorizerId is required'
                })
            };
        }

        const updated = [];
        const failed = [];

        // Process candidates in parallel to avoid timeout
        const processCandidate = async (candidate) => {
            const regId = candidate.registrationId;
            let pkValue = candidate.pk;
            let skValue = candidate.sk;

            try {
                const assignedAt = new Date().toISOString();

                // If pk/sk missing (old format fallback), we MUST scan to find them
                if (!pkValue || !skValue) {
                    // console.log(`⚠️ Missing pk/sk for ${regId}, performing fallback scan...`);
                    const scanCommand = new ScanCommand({
                        TableName: TABLE_NAME,
                        FilterExpression: 'registrationId = :regId',
                        ExpressionAttributeValues: { ':regId': { S: regId } },
                        Limit: 1
                    });
                    const scanResult = await client.send(scanCommand);
                    if (!scanResult.Items || scanResult.Items.length === 0) {
                        throw new Error('Candidate not found via scan');
                    }
                    pkValue = scanResult.Items[0].pk?.S;
                    skValue = scanResult.Items[0].sk?.S;
                }

                if (!pkValue || !skValue) {
                    throw new Error('Missing pk/sk keys');
                }

                // 1. Delete existing assignments (Query + Delete)
                const queryCommand = new QueryCommand({
                    TableName: CATEGORIZER_ASSIGNMENTS_TABLE,
                    KeyConditionExpression: 'pk = :pk',
                    ExpressionAttributeValues: { ':pk': regId }
                });

                const existingAssignments = await docClient.send(queryCommand);

                const deletePromises = [];
                if (existingAssignments.Items && existingAssignments.Items.length > 0) {
                    for (const oldAssignment of existingAssignments.Items) {
                        deletePromises.push(docClient.send(new DeleteCommand({
                            TableName: CATEGORIZER_ASSIGNMENTS_TABLE,
                            Key: { pk: oldAssignment.pk, sk: oldAssignment.sk }
                        })));
                    }
                    await Promise.all(deletePromises);
                }

                // 2. Create new assignment record & 3. Update main table (Parallel)
                await Promise.all([
                    docClient.send(new PutCommand({
                        TableName: CATEGORIZER_ASSIGNMENTS_TABLE,
                        Item: {
                            pk: regId,
                            sk: `CATEGORIZER#${categorizerId}`,
                            categorizerId: categorizerId,
                            categorizerName: categorizerName || categorizerId,
                            assignedAt: assignedAt,
                            status: 'pending'
                        }
                    })),
                    docClient.send(new UpdateCommand({
                        TableName: TABLE_NAME,
                        Key: { pk: pkValue, sk: skValue },
                        UpdateExpression: 'SET assignedCategorizer = :categorizer, categorizerAssignedAt = :assignedAt',
                        ExpressionAttributeValues: {
                            ':categorizer': categorizerId,
                            ':assignedAt': assignedAt
                        }
                    }))
                ]);

                return { success: true, regId };

            } catch (error) {
                console.error(`❌ Failed to assign ${regId}:`, error);
                return { success: false, regId, error: error.message };
            }
        };

        // Execute all assignments in parallel
        const results = await Promise.all(candidatesToProcess.map(candidate => processCandidate(candidate)));

        // Process results
        results.forEach(result => {
            if (result.success) {
                updated.push(result.regId);
            } else {
                failed.push({ registrationId: result.regId, error: result.error });
            }
        });

        console.log(`✅ Bulk assignment complete. Success: ${updated.length}, Failed: ${failed.length}`);

        return {
            statusCode: 200,
            headers: corsHeaders,
            body: JSON.stringify({
                success: true,
                message: `Assigned ${updated.length} candidates to ${categorizerName || categorizerId}`,
                updated: updated,
                failed: failed,
                summary: {
                    total: totalCount,
                    successful: updated.length,
                    failed: failed.length
                }
            })
        };

    } catch (error) {
        console.error('❌ Error in assign categorizer:', error);
        return {
            statusCode: 500,
            headers: corsHeaders,
            body: JSON.stringify({
                success: false,
                message: `Error assigning categorizers: ${error.message}`
            })
        };
    }
};

File: lambda-functions/assign-categorizer.py
import json
import boto3
from datetime import datetime

# Initialize DynamoDB client
dynamodb = boto3.resource('dynamodb')

# TODO: Replace with your actual table names
CANDIDATES_TABLE_NAME = 'your-candidates-table-name'  # Main candidates table
CATEGORIZER_ASSIGNMENTS_TABLE = 'categorizer-assignments'  # New table for categorizer assignments

candidates_table = dynamodb.Table(CANDIDATES_TABLE_NAME)
assignments_table = dynamodb.Table(CATEGORIZER_ASSIGNMENTS_TABLE)

def lambda_handler(event, context):
    """
    Lambda function to assign candidates to categorizers
    
    Expected payload:
    {
        "registrationIds": ["REG123", "REG124", "REG125"],
        "categorizerId": "CAT001",
        "categorizerName": "categorizer1"
    }
    """
    
    try:
        # Parse request body
        if isinstance(event.get('body'), str):
            body = json.loads(event['body'])
        else:
            body = event
        
        # Extract parameters
        registration_ids = body.get('registrationIds', [])
        categorizer_id = body.get('categorizerId')
        categorizer_name = body.get('categorizerName')
        
        # Validate required fields
        if not registration_ids or not isinstance(registration_ids, list):
            return {
                'statusCode': 400,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                    'Access-Control-Allow-Headers': 'Content-Type',
                    'Access-Control-Allow-Methods': 'POST, OPTIONS'
                },
                'body': json.dumps({
                    'success': False,
                    'message': 'registrationIds array is required'
                })
            }
        
        if not categorizer_id:
            return {
                'statusCode': 400,
                'headers': {
                    'Content-Type': 'application/json',
                    'Access-Control-Allow-Origin': '*',
                    'Access-Control-Allow-Headers': 'Content-Type',
                    'Access-Control-Allow-Methods': 'POST, OPTIONS'
                },
                'body': json.dumps({
                    'success': False,
                    'message': 'categorizerId is required'
                })
            }
        
        updated = []
        failed = []
        
        # Process each registration ID
        for reg_id in registration_ids:
            try:
                # Create assignment record in categorizer assignments table
                assignment_item = {
                    'pk': reg_id,
                    'sk': f'CATEGORIZER#{categorizer_id}',
                    'categorizerId': categorizer_id,
                    'categorizerName': categorizer_name or categorizer_id,
                    'assignedAt': datetime.utcnow().isoformat(),
                    'status': 'pending'  # pending, categorized
                }
                
                assignments_table.put_item(Item=assignment_item)
                
                # Also update the main candidates table with categorizer assignment
                candidates_table.update_item(
                    Key={
                        'pk': reg_id,
                        # 'sk': 'CANDIDATE'  # Add if you use sort key
                    },
                    UpdateExpression='SET assignedCategorizer = :categorizer, categorizerAssignedAt = :assignedAt',
                    ExpressionAttributeValues={
                        ':categorizer': categorizer_id,
                        ':assignedAt': datetime.utcnow().isoformat()
                    }
                )
                
                updated.append(reg_id)
                print(f"Successfully assigned {reg_id} to {categorizer_id}")
                
            except Exception as e:
                print(f"Failed to assign {reg_id}: {str(e)}")
                failed.append({
                    'registrationId': reg_id,
                    'error': str(e)
                })
        
        return {
            'statusCode': 200,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type',
                'Access-Control-Allow-Methods': 'POST, OPTIONS'
            },
            'body': json.dumps({
                'success': True,
                'message': f'Assigned {len(updated)} candidates to {categorizer_name or categorizer_id}',
                'updated': updated,
                'failed': failed,
                'summary': {
                    'total': len(registration_ids),
                    'successful': len(updated),
                    'failed': len(failed)
                }
            })
        }
        
    except Exception as e:
        print(f"Error in assign categorizer: {str(e)}")
        return {
            'statusCode': 500,
            'headers': {
                'Content-Type': 'application/json',
                'Access-Control-Allow-Origin': '*',
                'Access-Control-Allow-Headers': 'Content-Type',
                'Access-Control-Allow-Methods': 'POST, OPTIONS'
            },
            'body': json.dumps({
                'success': False,
                'message': f'Error assigning categorizers: {str(e)}'
            })
        }
